mirrord/agent/src/steal/connection.rs

```diff
@@ -0,0 +1,354 @@
+use streammap_ext::StreamMap;
+use tokio::{
+    io::{AsyncWriteExt, ReadHalf, WriteHalf},
+    net::TcpStream,
+};
+use tokio_stream::StreamExt;
+use tokio_util::io::ReaderStream;
+use tracing::error;
+
+use super::*;
+
+/// Created once per agent during initialization.
+///
+/// Runs as a separate thread while the agent lives.
+///
+/// - (agent -> stealer) communication is handled by [`command_rx`];
+/// - (stealer -> agent) communication is handled by [`client_senders`], and the [`Sender`] channels
+///   come inside [`StealerCommand`]s through  [`command_rx`];
+pub(crate) struct TcpConnectionStealer {
+    port_subscriptions: Subscriptions<Port, ClientID>,
+
+    /// Communication between (agent -> stealer) task.
+    ///
+    /// The agent controls the stealer task through [`TcpStealerAPI::command_tx`].
+    command_rx: Receiver<StealerCommand>,
+
+    /// Connected clients (layer instances) and the channels which the stealer task uses to send
+    /// back messages (stealer -> agent -> layer).
+    clients: HashMap<ClientID, Sender<DaemonTcp>>,
+    index_allocator: IndexAllocator<ConnectionId>,
+
+    /// Intercepts the connections, instead of letting them go through their normal pathways, this
+    /// is used to steal the traffic.
+    stealer: TcpListener,
+
+    /// Set of rules the agent uses to steal traffic from through the
+    /// [`TcpConnectionStealer::stealer`] listener.
+    iptables: SafeIpTables<iptables::IPTables>,
+
+    /// Used to send data back to the original remote connection.
+    write_streams: HashMap<ConnectionId, WriteHalf<TcpStream>>,
+
+    /// Used to read data from the remote connections.
+    read_streams: StreamMap<ConnectionId, ReaderStream<ReadHalf<TcpStream>>>,
+
+    /// Associates a `ConnectionId` with a `ClientID`, so we can send the data we read from
+    /// [`TcpConnectionStealer::read_streams`] to the appropriate client (layer).
+    client_connections: HashMap<ConnectionId, ClientID>,
+}
+
+impl TcpConnectionStealer {
+    /// Initializes a new [`TcpConnectionStealer`] fields, but doesn't start the actual working
+    /// task (call [`TcpConnectionStealer::start`] to do so).
+    #[tracing::instrument(level = "trace")]
+    pub(crate) async fn new(
+        command_rx: Receiver<StealerCommand>,
+        pid: Option<u64>,
+    ) -> Result<Self, AgentError> {
+        if let Some(pid) = pid {
+            let namespace = PathBuf::from("/proc")
+                .join(PathBuf::from(pid.to_string()))
+                .join(PathBuf::from("ns/net"));
+
+            set_namespace(namespace).unwrap();
+        }
+
+        Ok(Self {
+            port_subscriptions: Subscriptions::new(),
+            command_rx,
+            clients: HashMap::with_capacity(8),
+            index_allocator: IndexAllocator::new(),
+            stealer: TcpListener::bind((Ipv4Addr::UNSPECIFIED, 0)).await?,
+            iptables: SafeIpTables::new(iptables::new(false).unwrap())?,
```

Comment:
nitpick -if we could initialize iptables only when needed (if there's active steal) and remove when there are no steals, that'd be great (in terms of clean) - don't put much effort but if it's like 30-40min work then it's worth it

Reply:
https://github.com/meowjesty/mirrord/pull/5 Like this?